service/storage_service.cc

```diff
@@ -6202,7 +6202,10 @@ future<service::tablet_operation_repair_result> storage_service::repair_tablet(l
         if (trinfo->stage != locator::tablet_transition_stage::repair && trinfo->stage != locator::tablet_transition_stage::rebuild_repair) {
             throw std::runtime_error(fmt::format("Tablet {} stage is not at repair", tablet));
         }
-        if (trinfo->session_id) {
+        if (session_id && trinfo->session_id != session_id) {
+            throw std::runtime_error(fmt::format("Repair session_id = {} of tablet {} is outdated. {}", session_id, tablet,
+                trinfo->session_id ? fmt::format("Current session_id = {}", trinfo->session_id) : "No active session"));
+        } else if (trinfo->session_id) {
             slogger.debug("repair_tablet: tablet={} session_id={}", tablet, trinfo->session_id);
         } else {
             throw std::runtime_error(fmt::format("Tablet {} session is not set", tablet));
```

Comment:
nit: This maze of ifs works, but it could be simpler.

```
auto session = session_id ? session_id : trinfo->session_id;
```
And let the materialized guard fail if session is not active.